middleware/gzip.go

```diff
@@ -0,0 +1,324 @@
+//   (a) https://github.com/tmthrgd/gziphandler whose license(Apache License, Version 2.0) can be found here:                   https://github.com/tmthrgd/gziphandler/blob/9e3dc377f14f3554d9ae767761e33a87b38ed3f4/LICENSE.md
+//   (b) https://github.com/klauspost/compress/tree/master/gzhttp whose license(Apache License, Version 2.0) can be found here: https://github.com/klauspost/compress/blob/4bc73d36928c39bbd7cf823171081d14c884edde/gzhttp/LICENSE
+
+const (
+	acHeader           = "Accept-Encoding"
+	noCompression      = gzip.NoCompression
+	bestSpeed          = gzip.BestSpeed
+	bestCompression    = gzip.BestCompression
+	defaultCompression = gzip.DefaultCompression
+	huffmanOnly        = gzip.HuffmanOnly
+
+	defaultLevel = bestSpeed // TODO: should this be the default?
+
+	// TODO: copy/add docs.
+	// TODO: check other impls for other sizes that they use.
+	defaultMinSize = 150
+)
+
+// Gzip is a middleware that transparently gzips the response body, for clients which support.
+func Gzip(wrappedHandler http.HandlerFunc) http.HandlerFunc {
+
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		w.Header().Add(varyHeader, acHeader)
+
+		if !shouldGzip(r) {
+			wrappedHandler(w, r)
+			return
+		}
+
+		gw := &gzipRW{
+			ResponseWriter: w,
+			// h:              h,
+			buf: bufferPool.Get().(*[]byte),
+		}
+		defer func() {
+			err := gw.Close()
+			if err != nil && os.Getenv("GOWEB_RUNNING_IN_TESTS") != "" {
+				panic(err)
+			}
+		}()
+
+		var modifiedRw http.ResponseWriter = gw
+
+		_, cok := w.(http.CloseNotifier)
+		_, hok := w.(http.Hijacker)
+		_, pok := w.(http.Pusher)
+
+		_ = cok
+		_ = hok
+		_ = pok
+
+		// TODO: implement this
+		// switch {
+		// case cok && hok:
+		// 	modifiedRw = closeNotifyHijackResponseWriter{gw}
+		// case cok && pok:
+		// 	modifiedRw = closeNotifyPusherResponseWriter{gw}
+		// case cok:
+		// 	modifiedRw = closeNotifyResponseWriter{gw}
+		// case hok:
+		// 	modifiedRw = hijackResponseWriter{gw}
+		// case pok:
+		// 	modifiedRw = pusherResponseWriter{gw}
+		// }
+
+		wrappedHandler(modifiedRw, r)
+	}
+}
+
+var bufferPool = &sync.Pool{
+	New: func() interface{} {
+		buf := make([]byte, 0, defaultMinSize)
+		return &buf
+	},
+}
+
+var gzipWriterPools [gzip.BestCompression - gzip.HuffmanOnly + 1]sync.Pool
+
+func gzipWriterPool(level int) *sync.Pool {
+	return &gzipWriterPools[level-gzip.HuffmanOnly]
+}
+
+func gzipWriterGet(w io.Writer, level int) *gzip.Writer {
+	if gw, ok := gzipWriterPool(level).Get().(*gzip.Writer); ok {
+		gw.Reset(w)
+		return gw
+	}
+
+	gw, _ := gzip.NewWriterLevel(w, level)
+	return gw
+}
+
+func gzipWriterPut(gw *gzip.Writer, level int) {
+	gzipWriterPool(level).Put(gw)
+}
+
+// gzipRW provides an http.ResponseWriter interface,
+// which gzips bytes before writing them to the underlying
+// response. This doesn't close the writers, so don't forget
+// to do that. It can be configured to skip response smaller
+// than minSize.
+type gzipRW struct {
+	http.ResponseWriter
```

Comment:
do not wrap responseWriter since it breaks optional interfaces 
https://github.com/rs/cors/pull/84#discussion_r294039093

Reply:
covered; https://github.com/komuw/goweb/blob/976012c7578c83dceb46dd699113040eaa005d4f/middleware/gzip.go#L82-L91